src/transformStateWithClones.js

```diff
@@ -4,10 +4,32 @@
  * @param {Object} state
  * @param {Object[]} actions
  *
- * @return {Object[]}
+ * @return x
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentState = { ...state };
+  const states = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+      case 'removeProperties':
+        currentState = { ...currentState };
```

Comment:
[DONT REPEAT YOURSELF]: Since you are creating a new object copy in each case, you can do it just once before the switch statement to avoid repetition.